sky/shell/platform/ios/semantics_listener_impl.mm

```diff
@@ -0,0 +1,245 @@
+  assert(_nodes.find(node->id) != _nodes.end());
+  assert(_nodes.at(node->id)->parent == NULL);
+  _nodes.erase(node->id);
+  for (AccessibilityNode* child : node->children) {
+    removePersistentNode(child);
+  }
+}
+
+NSArray* SemanticsListenerImpl::createAccessibleElements() {
+  NSMutableArray* accessibleElements = [[NSMutableArray alloc] init];
+  for (NodeMapIterator iter = _nodes.begin(); iter != _nodes.end(); iter++) {
+    if (iter->second->parent == NULL) {
+      iter->second->populateAccessibleElements(accessibleElements);
+    }
+  }
+  return accessibleElements;
+}
+
+SemanticsListenerImpl::~SemanticsListenerImpl() {}
+
+SemanticsListenerImpl::AccessibilityNode::AccessibilityNode(
+    SemanticsListenerImpl* outer,
+    const semantics::SemanticsNodePtr& node)
+    : outer(outer) {
+  update(node);
+}
+
+void SemanticsListenerImpl::AccessibilityNode::update(
+    const semantics::SemanticsNodePtr& node) {
+  if (id == (uint32_t)-1) {
+    id = node->id;
+  }
+  assert(id == node->id);
+
+  if (!node->flags.is_null()) {
+    canBeTapped = node->flags->canBeTapped;
+    canBeLongPressed = node->flags->canBeLongPressed;
+    canBeScrolledHorizontally = node->flags->canBeScrolledHorizontally;
+    canBeScrolledVertically = node->flags->canBeScrolledVertically;
+    hasCheckedState = node->flags->hasCheckedState;
+    isChecked = node->flags->isChecked;
+  }
+
+  if (!node->strings.is_null()) {
+    if (label) {
+      delete label;
+    }
+    label = (char*)malloc((node->strings->label.size() + 1) * sizeof(char));
+    std::strcpy(label, node->strings->label.data());
+  }
+
+  if (!node->geometry.is_null()) {
+    if (!node->geometry->transform.is_null()) {
+      float* values = asFloatArray(node->geometry->transform.Pass());
+      transform.setColMajorf(values);
+      delete values;
+    }
+    left = node->geometry->left;
+    top = node->geometry->top;
+    width = node->geometry->width;
+    height = node->geometry->height;
+  }
+
+  if (!node->children.is_null()) {
+    const std::vector<AccessibilityNode*>& oldChildren = children;
+    children = std::vector<AccessibilityNode*>();
+    for (AccessibilityNode* child : oldChildren) {
+      assert(child->parent != NULL);
+      child->parent = NULL;
+      // TODO(tvolkert): recursively set parent to null on descendant
+      // tree, or assert may fail in removePersistentNode()
+    }
+    for (const semantics::SemanticsNodePtr& childNode : node->children) {
+      AccessibilityNode* child = outer->updateSemanticsNode(childNode);
+      child->parent = this;
+      children.push_back(child);
+    }
+    for (AccessibilityNode* child : oldChildren) {
+      if (child->parent == NULL) {
+        outer->removePersistentNode(child);
+      }
+    }
+  }
+
+  // Has to be done after children are updated
+  // since they also get marked dirty
+  invalidateGlobalGeometry();
+}
+
+float* asFloatArray(mojo::Array<float> input) {
```

Comment:
I see, so the caller is required to free() the resulting float*.  You might want to make that clear.  Better yet would be to use a uniqe_ptr.